cmd/kwil-admin/cmds/setup/genesis.go

```diff
@@ -0,0 +1,219 @@
+
+					uintHeight := uint64(height)
+					genesisInfo.ForkHeights[parts[1]] = &uintHeight
+				}
+			}
+			if cmd.Flags().Changed(chainIDFlag) {
+				if migration {
+					return makeErr(errors.New("cannot set chain ID when generating a migration genesis file"))
+				}
+
+				genesisInfo.ChainID = chainID
+			}
+			if cmd.Flags().Changed(genesisStateFlag) {
+				if migration {
+					return makeErr(errors.New("cannot set genesis state when generating a migration genesis file"))
+				}
+
+				apphash, err := appHashFromSnapshotFile(genesisState)
+				if err != nil {
+					return makeErr(err)
+				}
+
+				genesisInfo.DataAppHash = apphash
+			}
+			genesisInfo.ConsensusParams.WithoutGasCosts = !withGasCosts
+			if cmd.Flags().Changed(maxBytesPerBlockFlag) {
+				genesisInfo.ConsensusParams.Block.MaxBytes = maxBytesPerBlock
+			}
+			if cmd.Flags().Changed(joinExpiryFlag) {
+				genesisInfo.ConsensusParams.Validator.JoinExpiry = joinExpiry
+			}
+			if cmd.Flags().Changed(voteExpiryFlag) {
+				genesisInfo.ConsensusParams.Votes.VoteExpiry = voteExpiry
+			}
+			if cmd.Flags().Changed(maxVotesPerTxFlag) {
+				genesisInfo.ConsensusParams.Votes.MaxVotesPerTx = maxVotesPerBlock
+			}
+
+			if migration {
+				genesisInfo.Validators = nil
+				genesisInfo.Alloc = nil
+				genesisInfo.ForkHeights = nil
+			}
+
+			existingFile, err := os.Stat(out)
+			if err == nil && existingFile.IsDir() {
+				return makeErr(fmt.Errorf("a directory already exists at %s, please remove it first", out))
+			} else if err == nil {
+				return makeErr(fmt.Errorf("%s already exists, please remove it first", out))
+			}
+
+			err = genesisInfo.SaveAs(out)
+			if err != nil {
+				return makeErr(err)
+			}
+
+			return display.PrintCmd(cmd, display.RespString("Created genesis.json file at "+out))
+		},
+	}
+
+	cmd.Flags().StringVar(&output, outputFlag, "", "Output directory for the genesis.json file")
+	cmd.Flags().StringVar(&chainID, chainIDFlag, "", "Chain ID for the genesis.json file")
+	cmd.Flags().StringArrayVar(&validators, validatorsFlag, nil, "Public keys and power of initial validator(s)")
+	cmd.Flags().StringArrayVar(&allocs, allocsFlag, nil, "Address and initial balance allocation(s)")
+	cmd.Flags().StringArrayVar(&forks, forksFlag, nil, "Block height and name of fork(s)")
+	cmd.Flags().BoolVar(&withGasCosts, withGasCostsFlag, false, "Include gas costs in the genesis file")
+	cmd.Flags().Int64Var(&maxBytesPerBlock, maxBytesPerBlockFlag, 0, "Maximum number of bytes per block")
+	cmd.Flags().Int64Var(&joinExpiry, joinExpiryFlag, 0, "Number of blocks before a join proposal expires")
+	cmd.Flags().Int64Var(&voteExpiry, voteExpiryFlag, 0, "Number of blocks before a vote proposal expires")
+	cmd.Flags().Int64Var(&maxVotesPerBlock, maxVotesPerTxFlag, 0, "Maximum number of votes per validator transaction (each validator has 1 validator tx per block)")
+	cmd.Flags().BoolVar(&migration, migrationFlag, false, "Generate an incomplete genesis file for zero downtime migration")
```

Comment:
With **no** flags, the output genesis includes:

```json
    "migration": {
      "start_height": 0,
      "end_height": 0
    },
```

That's super confusing. I think we should modify the struct to omit empty unless setting the default empty to nil instead a poniter to zero value struct causes problems downstream somewhere.

Reply:
Ah good point